sql: use column IDs in composite secondary indexes

madelynnblue opened the pull request:
Change composite index encoding to always be preceeded by the column
ID diff. This allows both primary and secondary keys to use the same
decode path.

Since the column IDs are now known, datums can now also signal if
they need a composite encoding or not, since for many cases the key
encoding is reversible. Floats and decimals only do in the -0 case
or when zeros are truncated during key encoding. Collated strings
are not reversible and always need a composite encoding.

Float encodings have changed so 0 and -0 have the same key encodings
to allow for correct semantics during index operations. Floats have
been added to composite encodings so -0 can round trip in an index.

This is a backward-incompatible change. Any secondary indexes over
collated strings need to be dropped and recreated. They can be
dropped before or after upgrading, but must be recreated on the new
version. Failing to do this will result in an error: "pq: unknown
type NotNull" during a query.

Fixes #14345
See #14328
See 743842997a428cfae6805df114abf075d3b8129d

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14528)
<!-- Reviewable:end -->

mjibson commented:
I think this is backward incompatible with collated strings in secondary indexes, but I haven't been able to break anything yet.

RaduBerinde commented:
So decimals/floats are only double-encoded when it actually makes a difference, right? That's pretty cool.

For decimals it might still be beneficial to avoid encoding the full value again (vs just the number of 0s); I understand that is a much harder change to the encoding and decoding code, and I can't say for sure it would actually matter in practice, but perhaps it's worth filing an issue to investigate in the future.

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-KgarEzneSicdOwQt0N4:bw17idz) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
> ```Go
> 	}
> 	var x apd.Decimal
> 	x.Reduce(&d.Decimal)
> ```

Are we doing too much work here? We just need to know if `d.Coeff` ends in 0 anad the exponent is negative no?

---

*[pkg/sql/sqlbase/rowwriter.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgaqE67zUdQpGB2sGag:-KgaqE681_1aGKd_CJIo:b-7ixf4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/sqlbase/rowwriter.go#L96)):*
> ```Go
> // TODO(dan): This logic is common and being moved into TableDescriptor (see
> // #6233). Once it is, use the shared one.
> func (rh *rowHelper) skipColumnInPK(colID ColumnID, family FamilyID, value parser.Datum) (bool, error) {
> ```

This function needs a description now

---

*[pkg/sql/sqlbase/rowwriter.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-Kgaqin7rsSvpCuTgxJr:-Kgaqin7rsSvpCuTgxJs:b-3oy496) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/sqlbase/rowwriter.go#L120)):*
> ```Go
> 		return true, nil
> 	}
> 	return false, nil
> ```

This should be moved inside the if above (this case is the "else" to the IsComposite check). Better yet just `return !cdatm.IsComposite(), nil` there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-:-KgarVHb0JmuPgd3vlAS:bbymv41)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-Kgbhlq8TTSCc-hI8JlI:b-yj6etu) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are we doing too much work here? We just need to know if `d.Coeff` ends in 0 anad the exponent is negative no?
</blockquote></details>

We are maybe doing too much work here. https://github.com/cockroachdb/apd/pull/46 would help a bit because we could avoid the CmpTotal below, but the Reduce would still happen. That improvement could come later. I'd like to proceed as is and make some benchmarks for this in later commits.

---

*[pkg/sql/sqlbase/rowwriter.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgaqE67zUdQpGB2sGag:-KgbkeHzvXa-pBx-yRt1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/sqlbase/rowwriter.go#L96)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This function needs a description now
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-Kgaqin7rsSvpCuTgxJr:-KgbksB7Ck3RRwJTw-eP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/sqlbase/rowwriter.go#L120)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should be moved inside the if above (this case is the "else" to the IsComposite check). Better yet just `return !cdatm.IsComposite(), nil` there.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Backwards incompatibility at this point is unfortunate, but I guess it's OK since the feature has only been available for a couple of weeks. However, we should be sure that we have a suitable versioning mechanism in place so that if we had to make a change like this after 1.0 we could do it without invalidating existing indexes.

petermattis commented:
We do have `TableDescriptor.FormatVersion`, though I'm not sure if it would help maintain backward compatibility in this instance. @danhhz?

RaduBerinde commented:
*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-KgdoJZPYzwxd0hnvHER:b8z0dib) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

We are maybe doing too much work here. https://github.com/cockroachdb/apd/pull/46 would help a bit because we could avoid the CmpTotal below, but the Reduce would still happen. That improvement could come later. I'd like to proceed as is and make some benchmarks for this in later commits.
</blockquote></details>

Why do we need (the entire) Reduce? If we have 100 0s at the end, we don't need to figure out there are 100 of them, we just need to know if there is at least one.. Isn't this entire check equivalent to checking if the exponent is negative and `d.Coeff` is divisible by 10?

Ok with me to proceed as is though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
We could have preserved backward compatibility by having two different composite column ID lists, encoding the current one the way we do now and the new one with this new encoding. The engineering effort and ensuing permanent technical debt really don't seem worth it to me pre-1.0.

Minor comments below.

---

Reviewed 8 of 9 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/datum.go, line 493 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoJEW46q9QfeUbKo8O:-KgoJEW46q9QfeUbKo8P:b-2pfgn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/parser/datum.go#L493)):*
> ```Go
> func (d *DFloat) IsComposite() bool {
> 	// -0 is composite.
> 	return *d == 0 && math.Signbit(float64(*d))
> ```

Could also do `math.Float64bits(*d) == 1<<63`.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 302 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoK-ikAWjW_vx6SWyP:-KgoK-ikAWjW_vx6SWyQ:bk8tivk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/sqlbase/rowfetcher.go#L302)):*
> ```Go
> 	// undecodable bytes that came from the key so that there is no panic for
> 	// a duplicate value.
> 	for _, colID := range rf.index.CompositeColumnIDs {
> ```

I'm unclear on why this isn't guarded by the `familyID == 0` check any more. In primary indexes, all of the keys (hence all of the composite encodings) are in `familyID` 0. In secondary indexes, the `familyID` is always 0.

---

*[pkg/sql/sqlbase/rowwriter.go, line 305 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoLE9q6d4rY9P90WGp:-KgoLE9rsARyVK76qHbE:bq0ueay) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/sqlbase/rowwriter.go#L305)):*
> ```Go
> 				continue
> 			}
> 			if values[idx] == parser.DNull {
> ```

Combine this conditional `continue` with the previous one?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-:-KgoLqfx5DVv6sIb9nr5:btggoxo)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I've created #14575 to track an investigation to store the number of truncated zeros. As described in that issue, I'm skeptical it's worth it.

I've been able to demonstrate the backward compat issue. Creating a table with a secondary index on a collated key with the current beta and then selecting from that index on this version yields:

```
> SELECT * FROM t@t_a_key;
pq: unknown type NotNull
```

So not a catastrophic error. And easily fixable by dropping and recreating the index:

```
> drop index t@t_a_key;
DROP INDEX
root@:26257/d> CREATE UNIQUE INDEX ON t (a);
CREATE INDEX
root@:26257/d> SELECT * FROM t@t_a_key;
+---+
| a |
+---+
| a |
| A |
| b |
| B |
| ü |
| x |
+---+
(6 rows)
```

I've added this backward incompat description to the commit message.

---

Review status: 5 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-Kgoy58MTVqV5JdWvxgs:b-4em1xv) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why do we need (the entire) Reduce? If we have 100 0s at the end, we don't need to figure out there are 100 of them, we just need to know if there is at least one.. Isn't this entire check equivalent to checking if the exponent is negative and `d.Coeff` is divisible by 10?

Ok with me to proceed as is though.
</blockquote></details>

You're right, this was much simpler than I thought. Changed to check if d.Coeff is divisible by 10.

---

*[pkg/sql/parser/datum.go, line 493 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoJEW46q9QfeUbKo8O:-KgowuSr6oOwH_1LPT2Y:bha17ey) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/parser/datum.go#L493)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Could also do `math.Float64bits(*d) == 1<<63`.
</blockquote></details>

That's a copy of the implementation in Signbit, but it's less obvious what exactly it's doing. Especially with the go 1.9 inlining improvements I'd prefer to keep it as is because it should get compiled down to this same thing, but we keep the benefit of using a stable API.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 302 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoK-ikAWjW_vx6SWyP:-Kgp2fwadnKnPOiTf6y9:bhyom3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/sqlbase/rowfetcher.go#L302)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

I'm unclear on why this isn't guarded by the `familyID == 0` check any more. In primary indexes, all of the keys (hence all of the composite encodings) are in `familyID` 0. In secondary indexes, the `familyID` is always 0.
</blockquote></details>

I've moved this back to a familyID == 0 check, but that requires duplicating the loop because there's no familyID value in the secondary index case.

---

*[pkg/sql/sqlbase/rowwriter.go, line 305 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoLE9q6d4rY9P90WGp:-KgoyTo9r2Mg-KzP11NM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/sqlbase/rowwriter.go#L305)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Combine this conditional `continue` with the previous one?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-:-Kgp0es4-B5x8srZpFPq:bvjs4rg)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-Kgp7YkOB2SRTt8Ikw-W:b8tndau) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

You're right, this was much simpler than I thought. Changed to check if d.Coeff is divisible by 10.
</blockquote></details>

We also need to check the exponent no? `100` should not be composite (and `1.00` should).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/datum.go, line 493 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoJEW46q9QfeUbKo8O:-Kgp5sr7DstxMglUh6n1:b-vro9h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/parser/datum.go#L493)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

That's a copy of the implementation in Signbit, but it's less obvious what exactly it's doing. Especially with the go 1.9 inlining improvements I'd prefer to keep it as is because it should get compiled down to this same thing, but we keep the benefit of using a stable API.
</blockquote></details>

It's not just signbit; `return math.Float64bits(*d) == 1<<63` (or equivalently, `math.Float64bits(*d) == math.Float64bits(math.Copysign(0, -1))`) subsumes the comparison to zero as well, so I'm skeptical that the inliner would clean this up. IEEE 754 format and `math.Float64bits` are stable APIs.

I'm fine with the code as is though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-:-Kgp8BmIRSJPjA-dbdA8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-KgpE68vYrm5c2WnoOSz:bbz53ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We also need to check the exponent no? `100` should not be composite (and `1.00` should).
</blockquote></details>

I don't think so. `100` can be represented in an infinite number of ways: 1e2, 10e1, 100e0, 1000e-1, etc. In the 1e2 case the coefficient is `1`, so there's nothing to truncate. The encoding package converts the coefficient part to a string and truncates those zeros (during key encoding). The exponent is handled separately.

---

*[pkg/sql/parser/datum.go, line 493 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgoJEW46q9QfeUbKo8O:-KgpJA9_sDVPH2rEPzs0:b12g1dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43fe50d83a58c3d244a4d2f2a3f52f8b8c8c19c/pkg/sql/parser/datum.go#L493)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

It's not just signbit; `return math.Float64bits(*d) == 1<<63` (or equivalently, `math.Float64bits(*d) == math.Float64bits(math.Copysign(0, -1))`) subsumes the comparison to zero as well, so I'm skeptical that the inliner would clean this up. IEEE 754 format and `math.Float64bits` are stable APIs.

I'm fine with the code as is though.
</blockquote></details>

Oh I didn't realize it got rid of the 0 compare as well. Agreed, I've changed to use this suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-KgpJbgnEK5uy4ScB5SG:b-rjrg83) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I don't think so. `100` can be represented in an infinite number of ways: 1e2, 10e1, 100e0, 1000e-1, etc. In the 1e2 case the coefficient is `1`, so there's nothing to truncate. The encoding package converts the coefficient part to a string and truncates those zeros (during key encoding). The exponent is handled separately.
</blockquote></details>

We should avoid double-encoding all integers that end in `0`. If the exponent is non-negative, we absolutely don't need to double encode, and that should cover a lot of the cases even if it doesn't cover `1000e-1` no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: use column IDs in composite secondary indexes

Change composite index encoding to always be preceeded by the column
ID diff. This allows both primary and secondary keys to use the same
decode path.

Since the column IDs are now known, datums can now also signal if
they need a composite encoding or not, since for many cases the key
encoding is reversible. Floats and decimals only do in the -0 case
or when zeros are truncated during key encoding. Collated strings
are not reversible and always need a composite encoding.

Float encodings have changed so 0 and -0 have the same key encodings
to allow for correct semantics during index operations. Floats have
been added to composite encodings so -0 can round trip in an index.

This is a backward-incompatible change. Any secondary indexes over
collated strings need to be dropped and recreated. They can be
dropped before or after upgrading, but must be recreated on the new
version. Failing to do this will result in an error: "pq: unknown
type NotNull" during a query.

Fixes #14345
See #14328
See 743842997a428cfae6805df114abf075d3b8129d

mjibson commented:
Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528#-KgarEzneSicdOwQt0N3:-KgpYawBoighsWFcEbD4:by274f2) ([raw file](https://github.com/cockroachdb/cockroach/blob/423efcf037c6bf2671a0ba3457b1373daf1de4cc/pkg/sql/parser/datum.go#L623)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should avoid double-encoding all integers that end in `0`. If the exponent is non-negative, we absolutely don't need to double encode, and that should cover a lot of the cases even if it doesn't cover `1000e-1` no?
</blockquote></details>

We had a discussion in person and concluded that although we are not able today to find an example where preserving 0s on integers matters in terms of precision, it may in the future. Also it is easier to stop encoding these in the future if we are sure it's safe, but for now we are both ok with being conservative and always encoding them.

As an example of where this stuff can matter:

```
root@:26257/> SELECT pow(5::DECIMAL, 2);
+--------------------+
| pow(5::DECIMAL, 2) |
+--------------------+
|                 25 |
+--------------------+
(1 row)
root@:26257/> SELECT pow(5.0::DECIMAL, 2);
+----------------------+
| pow(5.0::DECIMAL, 2) |
+----------------------+
|                25.00 |
+----------------------+
```

However both of these involve zeros after the decimal, which isn't quite this problem, but it's the kind of behavior we want to preserve.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14528)*
<!-- Sent from Reviewable.io -->
